lib/linked_list.rb

```diff
@@ -14,73 +14,217 @@ def initialize(value, next_node = nil)
 class LinkedList
     def initialize
       @head = nil # keep the head private. Not accessible outside this class
+      @tail = nil
     end
 
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
     def add_first(value)
-      raise NotImplementedError
+      new_node = Node.new(value, nil)
+      new_node.next = @head
+      @head = new_node
     end
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
     def search(value)
-      raise NotImplementedError
+      if @head == nil
+        return false
+      end
+
+      current = @head
+      current_index = 0
+      while current != nil
+        if current.value == value
+          return true
+        else
+          current = current.next
+          current_index += 1
+        end
+      end
+
+      return false
     end
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max
-      raise NotImplementedError
+      if @head == nil
+        return nil
+      end
+
+      current = @head
+      current_index = 0
+      max_num = -1.0/0.0
+
+      while current != nil
+        if current.value > max_num
+          max_num = current.value
+        else 
+          current = current.next
+          current_index += 1
+        end

```

Comment:
I suggest instead of the `else` do this:

```ruby
end
current = current.next
```

You also don't need the current_index